### PR TITLE
Add support for Aquastream Ultimate sensors

### PR DESCRIPTION
**Pull request by aleksamagicka**
This adds support for Aquastream Ultimate sensors, reverse engineered and tested in #50.

### CODE REVIEW

**Review by leoratte — CHANGES_REQUESTED** [review]
Thanks for adding support for the Aquastream Ultimate.
This looks mostly good so far. The only issue I see at the moment is that the pump and fan readings are all implemented as special cases instead of using NUM_FANS.

**Comment by aleksamagicka** [comment]
Yes, that's next, I'll have to rearrange the labels to match, since the fan will take up the first slot.

**Comment by aleksamagicka** [comment]
I moved the fan to the `num_fans` method, the pump offsets are too all over the place for that. Also added what I think is the pump frequency, but that, as well as flow sensors, will have to be tested to iron out.

File: aquacomputer_d5next.c [hunk 1]
```diff
@@ -842,6 +916,11 @@ static umode_t aqc_is_visible(const void *data, enum hwmon_sensor_types type, u3
 	case hwmon_curr:
 		switch (priv->kind) {
 		case aquastreamxt:
+			/* Special case to support pump and fan current */
+			if (channel < 2)
+				return 0444;
+			break;
+		case aquastreamult:
```
Comment (leoratte) [inline aquacomputer_d5next.c:923]: You might have mixed up Aquastream XT and Ultimate here.

File: aquacomputer_d5next.c [hunk 2]
```diff
@@ -1958,6 +2066,8 @@ static int aqc_probe(struct hid_device *hdev, const struct hid_device_id *id)
 		priv->secondary_ctrl_report_id = AQUASTREAMXT_SECONDARY_CTRL_REPORT_ID;
 		priv->secondary_ctrl_report_size = AQUASTREAMXT_SECONDARY_CTRL_REPORT_SIZE;
 		priv->secondary_ctrl_report = aquastreamxt_secondary_ctrl_report;
+	} else if (priv->kind == aquastreamult) {
+		priv->fan_structure = &aqc_aquastreamult_fan_structure;
 	} else {
 		priv->serial_number_start_offset = AQC_SERIAL_START;
 		priv->firmware_version_offset = AQC_FIRMWARE_VERSION;
```
Comment (leoratte) [inline aquacomputer_d5next.c:2073]: I assume this won't work as the serial number and firmware version offsets are never set for the Aquastream Ultimate.

**Review by leoratte — CHANGES_REQUESTED** [review]
Thanks for the work. Please see my comments below.

File: aquacomputer_d5next.c (hunk 1 above)
Comment (aleksamagicka) [reply aquacomputer_d5next.c]: Embarrassing mistakes, thanks for noticing!

File: aquacomputer_d5next.c [hunk 3]
```diff
@@ -458,6 +482,35 @@ static const char *const label_aquastreamxt_temp_sensors[] = {
...
+	"Coolant temp",
+	"External temp"
+};
+
+static const char *const label_aquastreamult_speeds[] = {
+	"Fan speed",
+	"Pump speed",
+	"Pump frequency",
+	"Pressure [mbar]",
+	"Flow speed 1"
```
Comment (leoratte) [inline aquacomputer_d5next.c:496]: Same as above.

File: aquacomputer_d5next.c [hunk 4]
```diff
@@ -163,6 +166,27 @@ static u16 d5next_sensor_fan_offsets[] = { D5NEXT_PUMP_OFFSET, D5NEXT_FAN_OFFSET
...
+#define AQUASTREAMULT_COOLANT_TEMP		0x2D
+#define AQUASTREAMULT_EXTERNAL_TEMP		0x2F
+#define AQUASTREAMULT_PUMP_OFFSET		0x51
+#define AQUASTREAMULT_PUMP_VOLTAGE		0x3D
+#define AQUASTREAMULT_PUMP_CURRENT		0x53
+#define AQUASTREAMULT_PUMP_POWER		0x55
+#define AQUASTREAMULT_FAN_OFFSET		0x41
+#define AQUASTREAMULT_PUMP_FREQUENCY		0x15
+#define AQUASTREAMULT_PRESSURE_OFFSET		0x57
+#define AQUASTREAMULT_FLOW_1			0x37
```
Comment (leoratte) [inline aquacomputer_d5next.c:184]: Rename to 'AQUASTREAMULT_FLOW' as there is only one flow sensors.

File: aquacomputer_d5next.c [hunk 5]
```diff
@@ -1609,6 +1689,21 @@ static int aqc_raw_event(struct hid_device *hdev, struct hid_report *report, u8
 			i++;
 		}
 		break;
+	case aquastreamult:
+		/* Second temp sensor is not positioned after the first one, read it here */
+		priv->temp_input[1] = get_unaligned_be16(data + AQUASTREAMULT_EXTERNAL_TEMP) * 10;
```
Comment (leoratte) [inline aquacomputer_d5next.c:1694]: The addresses you provide above are directly after each other.

File: aquacomputer_d5next.c [hunk 6]
```diff
@@ -458,6 +482,35 @@ static const char *const label_aquastreamxt_temp_sensors[] = {
...
+	"Flow speed 1"
+};
+
+static const char *const label_aquastreamult_power[] = {
+	"Fan power",
+	"Pump power"
+};
+
+static const char *const label_aquastreamult_voltages[] = {
+	"Fan +12V",
```
Comment (leoratte) [inline aquacomputer_d5next.c:505]: The fan can be voltage controlled. so this is not always 12V.

**Review by leoratte — CHANGES_REQUESTED** [review]
I found some more small mistakes.

File: aquacomputer_d5next.c [hunk 7]
```diff
@@ -163,6 +166,27 @@ static u16 d5next_sensor_fan_offsets[] = { D5NEXT_PUMP_OFFSET, D5NEXT_FAN_OFFSET
...
 static u16 d5next_ctrl_fan_offsets[] = { 0x96, 0x41 };	/* Pump and fan speed (from 0-100%) */
 
+/* Specs of the Aquastream Ultimate pump */
+/* Pump does not follow the standard structure, so only consider the fan */
+#define AQUASTREAMULT_NUM_FANS		1
+#define AQUASTREAMULT_NUM_SENSORS	2
+
+/* Sensor report offsets for the Aquastream Ultimate pump */
+#define AQUASTREAMULT_COOLANT_TEMP		0x2D
+#define AQUASTREAMULT_EXTERNAL_TEMP		0x2F
```
Comment (leoratte) [inline aquacomputer_d5next.c:176]: Remove this as it is not used anymore.

Comment (aleksamagicka) [reply aquacomputer_d5next.c]: Thanks!

**Review by leoratte — APPROVED** [review]

**Comment by leoratte** [comment]
Should be ready to merge. Just add a short description of the Aquastream Ultimate capabilities to the docs.

**Comment by aleksamagicka** [comment]
Yeah, might as well do it here. Definitely need to clean it up a bit later.

**Comment by leoratte** [comment]
The aquastreamult_fan_sensor_offsets are missing causing the driver to crash.

**Pull request merged**